VerilogForVerificationStyle.md

```diff
@@ -0,0 +1,850 @@
+    <dut>_env_cfg.sv
+    <dut>_env_cov.sv
+    <dut>_scoreboard.sv
+    <dut>_virtual_sequencer.sv
+    seq_lib/
+      // These allow basic functionality to be added to the environment
+      <dut>_base_vseq.sv
+      <dut>_sanity_vseq.sv
+  tb/
+    tb.sv
+  tests/
+    <dut>_base_test.sv
+    <dut>_test_pkg.sv
+```
+* `<block>_agent_pkg.sv, <dut>_env_pkg.sv, <dut>_test_pkg.sv, and tb/tb.sv`
+must include these `` `include``-es and imports:
+
+```systemverilog
+`include "dv_macros.svh"
+`include "uvm_macros.svh"
+import uvm_pkg::*;
+```
+
+An example environment level package file for an arbitrary block `newblock` should look similar to this:
+
+```systemverilog
+package newblock_env_pkg;
+  // dep packages
+  import uvm_pkg::*;
+  import newblock_agent_pkg::*;
+
+  // macro includes
+  `include "uvm_macros.svh"
+  `include "dv_macros.svh"
+
+  // define any parameters
+  parameter int MY_BLOCK_PARAMETER = 10;
+
+  // define any local types/functions/tasks
+  typedef enum bit {
+    MyblockBooleanZero,
+    MyblockBooleanTrue
+  } myblock_boolean_e;
+
+  // Functions declared within packages must be automatic
+  function automatic bit myblock_test_function(myblock_boolean_e test_boolean);
+    //function definition
+  endfunction
+endpackage
+```
+
+## UVM Guidelines
+
+Always call `run_test()` without the test name argument. This allows command
+line test overrides to be performed using `+UVM_TESTNAME=<testname>` without
+having to recompile and allows testlists to be created for regression runs.
+
+
+### Class Definitions
+
+1.  After variable declarations, register any objects and components with the
+factory using the `uvm_component_utils` and `uvm_object_utils` macros.
+  *   Declare any member variables to be registered with the UVM factory field
+  automation macros `uvm_field_int, uvm_field_enum, uvm_field_object,
+  uvm_field_array_int`, etc. within `uvm_object/component_utils_begin` and
+  `uvm_object/component_utils_end`.
+2.  For all field automation macros, use `UVM_DEFAULT` as the flag argument.
+For example, if you want to enable everything except comparing and printing for
+a `rx_delay` field, the macro would look like:
+
+```systemverilog
+`uvm_field_int(rx_delay, UVM_DEFAULT | UVM_NOCOMPARE | UVM_NOPRINT)
+```
+
+3.  After factory registration, a constructor must be defined. Use the
+`uvm_object_new` macro for UVM objects and the `uvm_component_new` macro for UVM components.
+Both of these macros are defined in `dv_macros.svh`.
+4.  Any constructors written manually should contain a call to
+`super.new(name [,parent])` as the first line. Any other lines in the
+constructor are optional.
+5.  Do not add any extra arguments to the constructor other than `name` and `parent`.
+6.  All objects should generally be instantiated in the `build_phase()` function
+for components, or in the beginning of the `body()` task for sequences, unless
+  there is a good reason not to.
+  This is to enable proper usage of uvm_factory overrides.
```

Comment:
you might want to tick this: uvm_factory - > `uvm_factory`

Reply:
Yep, good catch!